R/dataRetrieval.R

```diff
@@ -0,0 +1,308 @@
+#' @importFrom SummarizedExperiment SummarizedExperiment
+#' @importFrom SummarizedExperiment rowData
+.create_se <- function(dataset = c("hpca", "blueprint_encode", "immgen", "mouse.rnaseq"),
+                       hub = ExperimentHub(), assays="normcounts",
+                       rm.NA = c("rows","cols","both","none"),
+                       has.rowdata=FALSE, has.coldata=TRUE) {
+    
+    ## TEMPORARY CODE for pulling data directly from github until it is 
+    ## available on ExHub =====================================================
+    ref.set <- getReferenceDataset(dataset=dataset)
+    nrmcnts <- ref.set$data
+    colnames(nrmcnts) <- paste(colnames(nrmcnts), 1:ncol(nrmcnts), sep = ".")
+    
+    ## handling NA's
+    if(!is.null(rm.NA)){
+        rm.NA <- match.arg(rm.NA)
+        nrmcnts <- .rm_NAs(nrmcnts, rm.NA)
+    }
+    
+    ## defining colDAta
+    coldata <- S4Vectors::DataFrame(row.names = colnames(nrmcnts),
+                                    label.main = ref.set$main_types,
+                                    label.fine = ref.set$types)
+    ref.se <- SummarizedExperiment(assays = list(normcounts = nrmcnts),
+                                   colData = coldata)
+    
+    return(ref.se)
+    
+    ##########################################################################
+    ## code for when data is actually available on ExpHub ===================
+    ##########################################################################
+    host <- file.path("SingleR", dataset)
+    
+    ## extract normalized values --------
+    all.assays <- list()
+    for (a in assays) {
+        nrmcnts <- hub[hub$rdatapath==file.path(host, sprintf("%s%s.rds", a, suffix))][[1]]
+        
+        if(!is.null(rm.NA)){
+            rm.NA <- match.arg(rm.NA)
+            nrmcnts <- .rm_NAs(nrmcnts, rm.NA)
+        }
+        all.assays[[a]] <- nrmcnts
+    }
+    
+    ## get metadata ----------------------
+    args <- list()
+    if (has.coldata) {
+        args$colData <- hub[hub$rdatapath==file.path(host, sprintf("coldata%s.rds", suffix))][[1]]
+    }
+    if (has.rowdata) {
+        args$rowData <- hub[hub$rdatapath==file.path(host, sprintf("rowdata%s.rds", suffix))][[1]]
+    }
+    
+    ## make the final SE object ----------
+    do.call(SummarizedExperiment, c(list(assays=all.assays), args))
+}
+
+
+#' Handle NAs in expression matrix
+#' 
+#' @param mat matrix of expression values
+#' @param rm.NA choices: "rows", "cols", "both"
+.rm_NAs <- function(mat, rm.NA = "rows"){
+    if(rm.NA == "rows" | rm.NA == "both"){
+        keep_rows <- apply(mat, 1, function(x) all(!is.na(x)))
+        if( length(keep_rows[keep_rows]) > 0){
+            mat <- mat[keep_rows, ]
+        }else{
+            warning("The number of rows (genes) with NAs equals the *total* number of genes. We're going to ignore your wish to eliminate rows with NAs.")
+        }
+        
+    }
+    
+    if(rm.NA == "cols" | rm.NA == "both"){
```

Comment:
`||` for scalars, `|` for vectors. The second term is not evaluated by `||` if the first term is true.